Add Awair sensor platform

ahayworth opened the pull request:
## Description:

This PR adds the Awair platform, to support [Awair](https://getawair.com) air quality monitors. A full description of the new platform may be found [here](https://github.com/home-assistant/home-assistant/commit/6ce81b5fad5b20834037e7c6f4eda1f4e545f713)

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#7579

## Example entry for `configuration.yaml` (if applicable):
```yaml
sensor:
  - platform: awair
    access_token: ACCESS_TOKEN
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

ahayworth committed:
Awair Sensor Platform

This commit adds a sensor platform for Awair devices, by accessing
their beta API. Awair heavily rate-limits this API, so we throttle
updates based on the number of devices found. We also allow for the
user to bypass API device listing entirely, because the device list
endpoint is limited to only 6 calls per day. A crashing or restarting
server would quickly hit that limit.

This sensor platform uses the python_awair library (also written
as part of this PR), which is available for async usage.

bachya commented:
The config helper addition should be moved into its own PR; when adding in tests, etc., both of these together a bit much for a single PR.

ahayworth commented:
> The config helper addition should be moved into its own PR; when adding in tests, etc., both of these together a bit much for a single PR.

@bachya - I'll submit that as a separate PR, but it's currently required for tests to pass. I'm not sure the best way to split that up without forcing these tests to fail (and thus confusing everyone involved).

ahayworth commented:
@bachya #18571 seemed like overkill to baloob, so I've closed it and re-factored this PR.

ahayworth committed:
Disable pylint warning for broad try/catch

It's true that this is generally not a great idea, but we really don't
want to crash here. If we can't set up the platform, logging it and
continuing is the right answer.

ahayworth committed:
Add space to satisfy the linter

ahayworth committed:
Awair platform PR feedback

- Bump python_awair to 0.0.2, which has support for more granular exceptions
- Ensure we have python_awair available in test
- Raise PlatformNotReady if we can't set up Awair
- Make the 'Awair score' its own sensor, rather than exposing it other ways
- Set the platform up as polling, and set a sensible default
- Pass in throttling parameters to the underlying data class, rather
than use hacky global variable access to dynamically set the interval
- Switch to dict access for required variables
- Use pytest coroutines, set up components via async_setup_component,
  and test/modify/assert in generally better ways
- Commit test data as fixtures

MartinHjelmare commented on homeassistant/components/sensor/awair.py:
```diff
@@ -148,57 +162,53 @@ def state(self):
+    @property
+    def should_poll(self):
+        """HA should poll us for state."""
+        return True
```
This is the default

ahayworth replied:
Thanks - that wasn't immediately obvious from the code. I'll remove it.

MartinHjelmare commented on homeassistant/components/sensor/awair.py:
```diff
@@ -148,57 +162,53 @@ def state(self):
         resp = await self._client.air_data_latest(self._uuid)
+
         timestamp = datetime.strptime(resp[0][ATTR_TIMESTAMP], TIME_FORMAT)
+        self.attrs[ATTR_TIMESTAMP] = timestamp
```
Is there a difference between this time and `last_updated` state attribute, where the latter attribute is when the state or state attributes last changed?

ahayworth replied:
It's the timestamp of the last time the device reported data to their servers. I think that's useful to have, because the API doesn't return connection state in any other way. I'll address the feedback though - it's clearly confusing, and I can make it less so. :)

MartinHjelmare commented on tests/components/sensor/test_awair.py:
```diff
@@ -0,0 +1,249 @@
+    sensor = hass.states.get('sensor.awair_score')
+    entity = hass.data[SENSOR_DOMAIN].get_entity('sensor.awair_score')
+    assert sensor.state == '78'
+    assert entity.device_class == DEVICE_CLASS_SCORE
```
These entity properties are also state attributes. Assert the state attributes instead. We want to avoid asserting things about the entities, since that makes the tests less robust.

ahayworth replied:
> These entity properties are also state attributes.
I was getting ready to quote the [documentation](https://dev-docs.home-assistant.io/en/master/api/core.html#homeassistant.core.State), but I played around with a debugger for a bit and realized that they're ... _attributes_. Not of the state object itself (which is how I interpreted your comment at first), but like `state.attributes['device_class']` attributes. :smile: 

I'll fix these - thanks. I was kinda sad about having to drop down into the entity!

MartinHjelmare commented on tests/components/sensor/test_awair.py:
```diff
@@ -0,0 +1,249 @@
+        # these sensors are marked as should_poll, so force the interval
+        for sensor in hass.states.async_all():
+            e_id = sensor.entity_id
+            await hass.helpers.entity_component.async_update_entity(e_id)
```
Instead patch `homeassistant.util.dt.utcnow` and use `async_fire_time_changed` callback from tests/common.py to fire a mock time event. Eg like this:
https://github.com/home-assistant/home-assistant/blob/dev/tests/components/alarm_control_panel/test_manual.py#L78-L82

Make sure to patch time when setting up the component too, since that's when the first update happens.

ahayworth replied:
:+1: I like `async_fire_time_changed`, that's elegant.

MartinHjelmare commented on tests/components/sensor/test_awair.py:
```diff
@@ -0,0 +1,249 @@
+        entity = hass.data[SENSOR_DOMAIN].get_entity('sensor.awair_score')
+
+        # pylint: disable=protected-access
+        await entity._data.async_update(no_throttle=True)
```
Remove this.

ahayworth committed:
Awair PR feedback, volume 2

- Don't force updates in test, instead modify time itself and let
  homeassistant update things "normally".
- Remove unneeded polling attribute
- Rename timestamp attribute to 'last_api_update', to better reflect
  that it is the timestamp of the last time the Awair API servers
  received data from this device.
- Use that attribute to flag the component as unavailable when data
  is stale. My own Awair device periodically goes offline and it really
  hardly indicates that at all.
- Dynamically set fixture timestamps to the test run utcnow() value,
  so that we don't have to worry about ancient timestamps in tests
  blowing up down the line.
- Don't assert on entities directly, for the most part. Find desired
  attributes in ... the attributes dict.

ahayworth commented:
@MartinHjelmare Okay - feedback addressed again. Let me know if there's anything else I can change to make it a better component. :)

ahayworth committed:
Merge branch 'dev' into awair2

ahayworth commented:
I really don't know why the tests are not passing remotely; they're passing locally. Perhaps my local test environment is messed up:
```
collecting ... 
 tests/components/sensor/test_awair.py ✓✓✓✓✓✓✓✓✓✓✓✓✓✓                                                                                                            100% ██████████
========================================================================== slowest 10 test durations ===========================================================================
0.02s teardown tests/components/sensor/test_awair.py::test_awair_score[pyloop]
0.02s teardown tests/components/sensor/test_awair.py::test_awair_temp[pyloop]
0.02s teardown tests/components/sensor/test_awair.py::test_awair_misc_attributes[pyloop]                                                                                       
0.02s teardown tests/components/sensor/test_awair.py::test_platform_automatically_configured[pyloop]                                                                           
0.02s teardown tests/components/sensor/test_awair.py::test_awair_co2[pyloop]
0.02s teardown tests/components/sensor/test_awair.py::test_awair_humid[pyloop]
0.02s teardown tests/components/sensor/test_awair.py::test_throttle_async_update[pyloop]                                                                                       
0.02s teardown tests/components/sensor/test_awair.py::test_awair_unsupported_sensors[pyloop]                                                                                   
0.02s teardown tests/components/sensor/test_awair.py::test_platform_manually_configured[pyloop]                                                                                
0.02s teardown tests/components/sensor/test_awair.py::test_awair_dust[pyloop]

Results (0.59s):
      14 passed
py37 runtests: commands[1] | /home/andrew/projects/home-assistant/script/check_dirty
___________________________________________________________________________________ summary ____________________________________________________________________________________
  py37: commands succeeded
  congratulations :)
[Thu Nov 22 17:43] andrew@alpha5:~/projects/home-assistant [awair u=]
```

ahayworth committed:
Patch an instance of utcnow I overlooked

MartinHjelmare commented on tests/components/sensor/test_awair.py:
```diff
@@ -0,0 +1,280 @@
+    future = NOW + timedelta(minutes=30)
+    data_patch = patch('python_awair.AwairClient.air_data_latest',
+                       return_value=mock_coro(AIR_DATA_FIXTURE))
+    time_patch = patch('homeassistant.util.utcnow',
```
Why are we patching this place? Are we using that path anywhere? I think we should patch `'homeassistant.util.dt.utcnow'`. That will patch the place that event helpers are using.

ahayworth replied:
Yes, it is needed. I swear it's not just adding code for the sheer fun of it. :smile: 

In [homeassistant/util/\_\_init.py\_\_](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/util/__init__.py#L17), we import `utcnow` like so:
```python
from .dt import as_local, utcnow
```

I think that python mocking is not as smart as we'd like to assume. I feel like the import path matters somehow - importing it at the base of `util` seems to change the import path. Regardless, it's better to just show, not tell. Here's with and without the patched method in question: https://gist.github.com/ahayworth/16bd4985c0105382d2247374d5e0f25e

MartinHjelmare replied:
Ok, I found where we use that path. It's in the `Throttle` class in `homeassistant.util`. So you're right that we need to patch that. But we should also patch the path that I requested above, since that is used during setup component.

MartinHjelmare commented on tests/components/sensor/test_awair.py:
```diff
@@ -0,0 +1,280 @@
+    if config is None:
+        config = DISCOVERY_CONFIG
+
+    with devices_patch, air_data_patch:
```
We should patch time here too with `NOW`. This is where the clock starts for the scheduled polling.

ahayworth committed:
Switch to using a context manager for timestream modification

Honestly, it's just a lot easier to keep track of patches. Moreover,
the ones I seem to have missed are now caught, and tests seem to
consistently pass.

Also, switch test_throttle_async_update to manipulating time more
explicitly.

ahayworth committed:
Missing blank line, thank you hound

ahayworth committed:
Fix pydocstyle error

I very much need to set up a script to do this quickly w/o tox, because
running flake8 is not enough!

MartinHjelmare commented on tests/components/sensor/test_awair.py:
```diff
@@ -0,0 +1,281 @@
+    with devices_patch, air_data_patch, alter_time(NOW):
+        assert await async_setup_component(hass, SENSOR_DOMAIN, config)
+
+    await hass.async_block_till_done()
```
Indent this under the context manager scope.

ahayworth replied:
:+1:

MartinHjelmare commented on tests/components/sensor/test_awair.py:
```diff
@@ -0,0 +1,281 @@
+    """Manage multiple time mocks."""
+    patch_one = patch('homeassistant.util.dt.utcnow', return_value=retval)
+    patch_two = patch('homeassistant.util.utcnow', return_value=retval)
+
```
Add a third time patch here for `'homeassistant.components.sensor.awair.dt.utcnow'`.

ahayworth replied:
:+1:

ahayworth committed:
PR feedback

MartinHjelmare commented:
Yeah, that test is flaky. We can ignore it.

MartinHjelmare reviewed (commented):
Looks good! Just some small comments left.

ahayworth committed:
PR feedback

ahayworth committed:
Merge branch 'dev' into awair

MartinHjelmare reviewed (approved):
Great! Can be merged when build passes.

ahayworth commented:
> Great! Can be merged when build passes.

Excellent, thank you! :)
